File: backend/models/asset.go
package models

import (
    "time"
    "gorm.io/gorm"
)

type AssetCategory struct {
    ID          uint           `json:"id" gorm:"primaryKey"`
    Code        string         `json:"code" gorm:"unique;not null;size:20"`
    Name        string         `json:"name" gorm:"not null;size:100"`
    Description string         `json:"description" gorm:"type:text"`
    ParentID    *uint          `json:"parent_id" gorm:"index"`
    IsActive    bool           `json:"is_active" gorm:"default:true"`
    CreatedAt   time.Time      `json:"created_at"`
    UpdatedAt   time.Time      `json:"updated_at"`
    DeletedAt   gorm.DeletedAt `json:"-" gorm:"index"`

    // Relations
    Parent   *AssetCategory `json:"parent,omitempty" gorm:"foreignKey:ParentID"`
    Children []AssetCategory `json:"children,omitempty" gorm:"foreignKey:ParentID"`
    Assets   []Asset         `json:"-" gorm:"foreignKey:CategoryID"`
}

type Asset struct {
    ID            uint           `json:"id" gorm:"primaryKey"`
    Code          string         `json:"code" gorm:"unique;not null;size:20"`
    Name          string         `json:"name" gorm:"not null;size:100"`
    CategoryID    *uint          `json:"category_id" gorm:"index"`
    Category      string         `json:"category" gorm:"size:50"` // Keep for backward compatibility
    Status        string         `json:"status" gorm:"not null;size:20"` // ACTIVE, INACTIVE, SOLD
    PurchaseDate  time.Time      `json:"purchase_date"`
    PurchasePrice float64        `json:"purchase_price" gorm:"type:decimal(15,2);default:0"`
    SalvageValue  float64        `json:"salvage_value" gorm:"type:decimal(15,2);default:0"`
    UsefulLife    int            `json:"useful_life" gorm:"default:0"` // Years
    DepreciationMethod string    `json:"depreciation_method" gorm:"size:20"` // STRAIGHT_LINE, DECLINING_BALANCE
    IsActive      bool           `json:"is_active" gorm:"default:true"`
    Notes         string         `json:"notes" gorm:"type:text"`
    Location      string         `json:"location" gorm:"size:100"`
    Coordinates   string         `json:"coordinates" gorm:"size:50"` // "lat,lng" format
    MapsURL       string         `json:"maps_url" gorm:"size:500"`  // Generated Google Maps URL
    SerialNumber  string         `json:"serial_number" gorm:"size:50"`
    Condition     string         `json:"condition" gorm:"size:20;default:'Good'"`
    ImagePath     string         `json:"image_path" gorm:"size:255"`
    CreatedAt     time.Time      `json:"created_at"`
    UpdatedAt     time.Time      `json:"updated_at"`
    DeletedAt     gorm.DeletedAt `json:"-" gorm:"index"`

    // Relations
    AssetAccountID     *uint     `json:"asset_account_id" gorm:"index"`
    DepreciationAccountID *uint  `json:"depreciation_account_id" gorm:"index"`
    AccumulatedDepreciation float64 `json:"accumulated_depreciation" gorm:"type:decimal(15,2);default:0"`
    AssetCategory   *AssetCategory `json:"asset_category,omitempty" gorm:"foreignKey:CategoryID"`
    AssetAccount    *Account     `json:"asset_account" gorm:"foreignKey:AssetAccountID"`
    DepreciationAccount *Account `json:"depreciation_account" gorm:"foreignKey:DepreciationAccountID"`
}

// Asset Status Constants
const (
    AssetStatusActive   = "ACTIVE"
    AssetStatusInactive = "INACTIVE"
    AssetStatusSold     = "SOLD"
)

// Depreciation Methods Constants
const (
    DepreciationMethodStraightLine       = "STRAIGHT_LINE"
    DepreciationMethodDecliningBalance   = "DECLINING_BALANCE"
)

File: frontend/src/types/asset.ts
export interface Asset {
  id: number;
  code: string;
  name: string;
  category: string;
  status: 'ACTIVE' | 'INACTIVE' | 'SOLD';
  purchase_date: string;
  purchase_price: number;
  salvage_value: number;
  useful_life: number;
  depreciation_method: 'STRAIGHT_LINE' | 'DECLINING_BALANCE';
  accumulated_depreciation: number;
  is_active: boolean;
  notes: string;
  location?: string;
  coordinates?: string;         // "lat,lng" format
  maps_url?: string;           // Generated Google Maps URL
  serial_number?: string;
  condition?: string;
  image_path?: string;          // Path to asset image
  asset_account_id?: number;
  depreciation_account_id?: number;
  created_at: string;
  updated_at: string;
  
  // Relations (if preloaded)
  asset_account?: {
    id: number;
    code: string;
    name: string;
  };
  depreciation_account?: {
    id: number;
    code: string;
    name: string;
  };
}

export interface AssetFormData {
  code?: string;
  name: string;
  category: string;
  status?: 'ACTIVE' | 'INACTIVE' | 'SOLD';
  purchaseDate: string;
  purchasePrice: number;
  salvageValue?: number;
  usefulLife: number;
  depreciationMethod?: 'STRAIGHT_LINE' | 'DECLINING_BALANCE';
  isActive?: boolean;
  notes?: string;
  location?: string;
  coordinates?: string;         // "lat,lng" format for maps
  serialNumber?: string;
  condition?: string;
  assetAccountId?: number;
  depreciationAccountId?: number;
  userId?: number;
}

export interface AssetsSummary {
  total_assets: number;
  active_assets: number;
  total_value: number;
  total_depreciation: number;
  net_book_value: number;
}

export interface DepreciationEntry {
  year: number;
  date: string;
  depreciation_cost: number;
  accumulated_depreciation: number;
  book_value: number;
}

export interface AssetDepreciationReport {
  asset: Asset;
  annual_depreciation: number;
  monthly_depreciation: number;
  remaining_depreciation: number;
  remaining_years: number;
  current_book_value: number;
}

export interface DepreciationCalculation {
  asset_id: number;
  asset_name: string;
  as_of_date: string;
  purchase_price: number;
  salvage_value: number;
  accumulated_depreciation: number;
  current_book_value: number;
  depreciation_method: string;
  useful_life_years: number;
}

// Account interface for dropdowns
export interface Account {
  id: number;
  code: string;
  name: string;
  type: 'ASSET' | 'LIABILITY' | 'EQUITY' | 'REVENUE' | 'EXPENSE';
  category: string;
  balance: number;
  is_active: boolean;
}

export const ASSET_CATEGORIES = [
  'Fixed Asset',
  'Real Estate',
  'Computer Equipment',
  'Vehicle',
  'Office Equipment',
  'Furniture',
  'IT Infrastructure',
  'Machinery'
] as const;

export const ASSET_STATUS = [
  'ACTIVE',
  'INACTIVE',
  'SOLD'
] as const;

export const DEPRECIATION_METHODS = [
  'STRAIGHT_LINE',
  'DECLINING_BALANCE'
] as const;

export const DEPRECIATION_METHOD_LABELS = {
  'STRAIGHT_LINE': 'Straight Line',
  'DECLINING_BALANCE': 'Declining Balance'
} as const;

export type AssetCategory = typeof ASSET_CATEGORIES[number];
export type AssetStatus = typeof ASSET_STATUS[number];
export type DepreciationMethod = typeof DEPRECIATION_METHODS[number];
